test-framework/core/src/main/java/org/keycloak/testframework/admin/KeycloakAdminClientFactory.java

```diff
@@ -0,0 +1,81 @@
+package org.keycloak.testframework.admin;
+
+import org.keycloak.OAuth2Constants;
+import org.keycloak.admin.client.Keycloak;
+import org.keycloak.admin.client.KeycloakBuilder;
+import org.keycloak.admin.client.resource.RealmResource;
+import org.keycloak.representations.idm.ClientRepresentation;
+import org.keycloak.representations.idm.UserRepresentation;
+import org.keycloak.testframework.TestFrameworkException;
+import org.keycloak.testframework.annotations.InjectAdminClientFactory;
+import org.keycloak.testframework.injection.InstanceContext;
+import org.keycloak.testframework.realm.ManagedRealm;
+
+import java.util.ArrayList;
+import java.util.List;
+
+
+public class KeycloakAdminClientFactory {
+
+    private final List<Keycloak> adminClients = new ArrayList<>();
+
+    private final InstanceContext<KeycloakAdminClientFactory, InjectAdminClientFactory> instanceContext;
+    private final String serverUrl;
+
+    public KeycloakAdminClientFactory(InstanceContext<KeycloakAdminClientFactory, InjectAdminClientFactory> instanceContext, String serverUrl) {
+        this.instanceContext = instanceContext;
+        this.serverUrl = serverUrl;
+    }
+
+    public Keycloak create(String realmRef, String clientId, String user, String password) {
+        ManagedRealm managedRealm = instanceContext.getDependency(ManagedRealm.class, realmRef);
+
+        KeycloakBuilder clientBuilder = createBuilder(managedRealm.getName());
+        clientBuilder.grantType(OAuth2Constants.PASSWORD);
+        RealmResource realmRes = managedRealm.admin();
+
+        setClient(clientBuilder, realmRes, clientId);
+        setUser(clientBuilder, realmRes, user, password);
+
+        Keycloak adminClient = clientBuilder.build();
+        adminClients.add(adminClient);
+        return adminClient;
+    }
+
+    public Keycloak create(String realmRef, String clientId) {
+        ManagedRealm managedRealm = instanceContext.getDependency(ManagedRealm.class, realmRef);
```

Comment:
@vaceksimon @stianst shouldn't it be rather in the supplier? I mean, the instance context and ManagedRealm dependency. I know it's not a strict rule, but we try for better visibility to keep it in a supplier.

Reply:
Yeah, instead of this, I'm going to have the `create` methods have a `realm`, not `realmRef` as an argument. So we don't create realms dynamically.